File: client/src/app/models/graph-data.ts
export interface Ownership {
    parentId: string;
    percentage: number;
}

export interface AddNodeData {
    label: string;
    ownershipData: Ownership[];
}

export interface Node {
    id: string;
    label: string;
    ownerships: Ownership[];
}

export interface Edge {
    id: string;
    sourceId: string;
    targetId: string;
}

export interface GraphData {
    id: string;
    graphName: string;
    layout: LayoutType;
    nodes: Node[];
    edges: Edge[];
}

export enum LayoutType {
    HierarchicLayout = 0,
    OrganicLayout = 1,
    OrthogonalLayout = 2,
    CircularLayout = 3,
    RadialLayout = 4
}

File: api/Models/GraphData.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace api.Models;
public class Node
{
    public required string Id { get; set; }
    public required string Label { get; set; }
    public List<Ownership> Ownerships { get; set; }
}
public class Edge
{
    public required string Id { get; set; }
    public required string SourceId { get; set; }
    public required string TargetId { get; set; }
}
public class Ownership
{
    public string ParentId { get; set; }
    public double Percentage { get; set; }
}
public class AddNodeData
{
    public string Label { get; set; }
    public List<Ownership> OwnershipData { get; set; }

    public AddNodeData()
    {
        Label = "";
        OwnershipData = new List<Ownership>();
    }
}
public class GraphData
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? Id { get; set; }
    public required List<Node> Nodes { get; set; }
    public required List<Edge> Edges { get; set; }
}

